### PR TITLE
Circle

### PR DESCRIPTION
**Pull request by vdutor**

### CODE REVIEW

File: geometric_kernels/eigenfunctions.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+            .. note:
+                The length of `weights` is equal to the number of levels.
+                This is **not** the same as the number of eigenfunctions.
+
+        :param X: Inputs where to evaluate the eigenfunctions, shape = [N, D].
+        :param X2: Inputs where to evaluate the eigenfunctions, shape = [N2, D].
+            Default to None, in which X is used for X2.
+        :return: shape [N, N2]
+        """
+        if X2 is None:
```
Comment (vabor112) [inline geometric_kernels/eigenfunctions.py:86]: Couldn't we avoid code duplication by putting `if X2 is None: X2=X`?

File: geometric_kernels/spaces/circle.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+from geometric_kernels.eigenfunctions import EigenfunctionWithAdditionTheorem
+from geometric_kernels.spaces import SpaceWithEigenDecomposition
+from geometric_kernels.types import TensorLike
+
+
+def cartesian_to_polar(X: TensorLike) -> TensorLike:
+    """
+    :param X: Cartesian coordinates on the circle, [N, 2].
+        This means their norm equals one.
+    :return: angle [N, 1]
```
Comment (vabor112) [inline geometric_kernels/spaces/circle.py:21]: Maybe put "angle (between PI and -PI) for clarity"?

File: geometric_kernels/spaces/circle.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+    @property
+    def num_eigenfunctions_per_level(self) -> np.ndarray:
+        """Number of eigenfunctions per level, [N_l]_{l=0}^{L-1}"""
+        return np.array(
+            [1 if level == 0 else 2 for level in range(self.num_levels)]
+        )
+
+
+class Circle(SpaceWithEigenDecomposition, gs.geometry.hypersphere.Hypersphere):
```
Comment (vabor112) [inline geometric_kernels/spaces/circle.py:124]: I like `DiscreteSpectrumSpace` a bit better than `SpaceWithEigenDecomposition`: it's a bit shorter and the spectrum is something closer to the notion of a geometric space than eigen decomposition is (also this is all debatable).
Btw, hope that I got the idea right and this is actually the different name for the `Space` for which we can provide Laplace--Beltrami eigenpairs.

File: geometric_kernels/spaces/circle.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        ), "num_eigenfunctions needs to be odd to include all eigenfunctions within a level."
+        assert num_eigenfunctions >= 1
+
+        self._num_eigenfunctions = num_eigenfunctions
+        # We know `num_eigenfunctions` is odd, therefore:
+        self._num_levels = num_eigenfunctions // 2 + 1
+
+    def __call__(self, X: TensorLike) -> TensorLike:
+        """
+        :param X: cartesian coordinates on the circle, [N, 2].
```
Comment (lrozo) [inline geometric_kernels/spaces/circle.py:44]: What does `N` correspond to?

File: geometric_kernels/spaces/circle.py (hunk 3 above)
Comment (lrozo) [reply geometric_kernels/spaces/circle.py:124]: @vabor112 your comment is actually very interesting. I am lacking a bit of background here, may you please point me out to some paper or source where I can check the differences you comment here?

File: geometric_kernels/spaces/circle.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+class Circle(SpaceWithEigenDecomposition, gs.geometry.hypersphere.Hypersphere):
+    r"""
+    Circle :math:`\mathbb{S}^1` manifold with sinusoids and cosines eigenfunctions.
+    """
+
+    def __init__(self):
+        super().__init__(dim=1)
+
+    def is_tangent(
```
Comment (lrozo) [inline geometric_kernels/spaces/circle.py:132]: Is this supposed to use the `geomstats` functionality to compute the tangent space?

**Review by lrozo — APPROVED** [review]
@vdutor this is great work, I like the classes structure and abstraction very much! Nice work! You will find several comments in my PR review, but nothing critical.

File: geometric_kernels/eigenfunctions.py (hunk 1 above)
Comment (vdutor) [reply geometric_kernels/eigenfunctions.py:110]: good catch

File: geometric_kernels/eigenfunctions.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+    @abc.abstractproperty
+    def num_eigenfunctions(self) -> int:
+        """Number of eigenfunctions, M"""
+        raise NotImplementedError
+
+
+class EigenfunctionWithAdditionTheorem(Eigenfunctions):
+    """
+    Eigenfunctions for which the sum over a level has a simpler expression.
```
[the remark this answers is not in the document]
Comment (vdutor) [reply geometric_kernels/eigenfunctions.py:74]: Added a better description with examples

**Pull request merged**